### PR TITLE
rgw: bucket layout fixes

### PR DESCRIPTION
**Pull request by cbodley**
a couple followup fixes for https://github.com/ceph/ceph/pull/34096

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by smanjara** [comment]
@cbodley, I cherry picked the commits on top of my local branch to test, and I noticed that it creates buckets with num_shards = 11. Will check further and update.

File: src/rgw/rgw_json_enc.cc [hunk 1]
```diff
@@ -818,7 +818,7 @@ void RGWBucketInfo::decode_json(JSONObj *obj) {
   JSONDecoder::decode_json("num_shards", layout.current_index.layout.normal.num_shards, obj);
   uint32_t hash_type;
   JSONDecoder::decode_json("bi_shard_hash_type", hash_type, obj);
-  layout.current_index.layout.normal.num_shards = (uint8_t)hash_type;
+  layout.current_index.layout.normal.hash_type = static_cast<rgw::BucketHashType>(hash_type);
```
Comment (mattbenjamin) [inline src/rgw/rgw_json_enc.cc:821]: ouch

File: src/rgw/services/svc_bi_rados.cc [hunk 2]
```diff
@@ -200,12 +200,12 @@ void RGWSI_BucketIndex_RADOS::get_bucket_index_object(const string& bucket_oid_b
 }
 
 int RGWSI_BucketIndex_RADOS::get_bucket_index_object(const string& bucket_oid_base, const string& obj_key,
-                                                     uint32_t num_shards, RGWBucketInfo::BIShardsHashType hash_type,
+                                                     uint32_t num_shards, rgw::BucketHashType hash_type,
                                                      string *bucket_obj, int *shard_id)
 {
   int r = 0;
   switch (hash_type) {
-    case RGWBucketInfo::MOD:
```
Comment (mattbenjamin) [inline src/rgw/services/svc_bi_rados.cc:208]: ew

**Comment by cbodley** [comment]
mostly clean run in http://pulpito.ceph.com/cbodley-2020-04-22_15:04:47-rgw-wip-cbodley-testing-distro-basic-smithi/

**Pull request merged**